pupil_src/shared_modules/surface_tracker/surface_tracker.py

```diff
@@ -0,0 +1,458 @@
+                fixations_on_surf = surface.map_events(
+                    fixation_events, self.camera_model
+                )
+
+                surface_event = {
+                    "topic": "surfaces.{}".format(surface.name),
+                    "name": surface.name,
+                    "surf_to_img_trans": surface.surf_to_img_trans.tolist(),
+                    "img_to_surf_trans": surface.img_to_surf_trans.tolist(),
+                    "gaze_on_surf": gaze_on_surf,
+                    "fixations_on_surf": fixations_on_surf,
+                    "timestamp": frame.timestamp,
+                }
+                events["surfaces"].append(surface_event)
+
+    @abstractmethod
+    def _update_surface_heatmaps(self):
+        pass
+
+    def gl_display(self):
+        self.gui.update()
+
+    def add_surface(self, _=None, init_dict=None):
+        if self.markers or init_dict is not None:
+            surface = self.Surface_Class(
+                name="Surface {:}".format(len(self.surfaces) + 1), init_dict=init_dict
+            )
+            self.surfaces.append(surface)
+            self.gui.add_surface(surface)
+            self._update_ui()
+        else:
+            logger.warning("Can not add a new surface: No markers found in the image!")
+
+    def remove_surface(self, i):
+        self.gui.remove_surface(self.surfaces[i])
+        del self.surfaces[i]
+        self._update_ui()
+        self.save_surface_definitions_to_file()
+
+    def on_notify(self, notification):
+        if notification["subject"] == "surface_tracker.surfaces_changed":
+            logger.info("Surfaces changed. Saving to file.")
+            self.save_surface_definitions_to_file()
+        elif notification["subject"].startswith(
+            "surface_tracker.heatmap_params_changed"
+        ):
+            self.save_surface_definitions_to_file()
+        elif notification["subject"].startswith("surface_tracker.surface_name_changed"):
+            self.save_surface_definitions_to_file()
+            self._update_ui()
+
+    def on_pos(self, pos):
+        self._last_mouse_pos = np.array(pos, dtype=np.float32)
+
+    def on_click(self, pos, button, action):
+        self.gui.on_click(pos, button, action)
+
+    def get_init_dict(self):
+        return {
+            "marker_min_perimeter": self.marker_min_perimeter,
+            "inverted_markers": self.inverted_markers,
+        }
+
+    def save_surface_definitions_to_file(self):
+        surface_definitions = file_methods.Persistent_Dict(
+            os.path.join(self._save_dir, "surface_definitions")
+        )
+        surface_definitions["surfaces"] = [
+            surface.save_to_dict() for surface in self.surfaces if surface.defined
+        ]
+        surface_definitions.save()
+
+    def deinit_ui(self):
+        self.g_pool.quickbar.remove(self.add_button)
+        self.add_button = None
+        self.remove_menu()
+
+    def cleanup(self):
+        self.save_surface_definitions_to_file()
```

Comment:
After reading the whole class, I think one could split it up into several parts, e.g.:

- The actual plugin (mostly holds and calls references to the other classes)
- General menu UI
- UI for individual surfaces
- Logic (updating surfaces etc)
- Persistence (loading / storing surfaces)

And perhaps more.

This would be quite some additional work, but currently you have to thoughtfully read the file from top to bottom to understand what all its tasks are (even after this review, I don't have a very clear picture of this in my head)

Reply:
I suggest we postpone this refactor until we need to extend surface tracking.